### ISSUE TITLE
High memory usage, possible memory leak

### ISSUE THREAD

**Issue by remicollet** [nbs-system/snuffleupagus#160]
When this extension is enabled, I encounter lot of memory issues which doesn't appear when disabled

Exemple, running twig test suite.

Without:

```
PHPUnit 5.7.27 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.1.15
Configuration: /dev/shm/BUILD/Twig-ffda54d85cc3aec3ba499c8c0eda9b2176edfe83/phpunit.xml

.............................................................   61 / 1352 (  4%)
<snip>
..........                                                    1352 / 1352 (100%)

Time: 788 ms, Memory: 34.00MB
```

With the extension enabled

```
PHPUnit 5.7.27 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.1.15
Configuration: /dev/shm/BUILD/Twig-ffda54d85cc3aec3ba499c8c0eda9b2176edfe83/phpunit.xml

.............................................................   61 / 1352 (  4%)
<snip>
.............................................................  427 / 1352 ( 31%)
.......................PHP Fatal error:  Allowed memory size of 134217728 bytes exhausted (tried to allocate 4096 bytes) in /dev/shm/BUILDROOT/php-twig-1.35.1-1.fc27.remi.7.1.x86_64/usr/share/php/Twig/Compiler.php on line 126

```

P.S. twig test suite is only an example, this happens with lot of other projects

**Comment by jvoisin** [comment]
On 03/03/2018 06:17, Remi Collet wrote:
> When this extension is enabled, I encounter lot of memory issue which doesn't appear when disable
> 
> Exemple, running twig test suite.
> 
> Without:
> 
> ```
> PHPUnit 5.7.27 by Sebastian Bergmann and contributors.
> 
> Runtime:       PHP 7.1.15
> Configuration: /dev/shm/BUILD/Twig-ffda54d85cc3aec3ba499c8c0eda9b2176edfe83/phpunit.xml
> 
> .............................................................   61 / 1352 (  4%)
> <snip>
> ..........                                                    1352 / 1352 (100%)
> 
> Time: 788 ms, Memory: 34.00MB
> ```
> 
> With the extension enabled
> 
> ```
> PHPUnit 5.7.27 by Sebastian Bergmann and contributors.
> 
> Runtime:       PHP 7.1.15
> Configuration: /dev/shm/BUILD/Twig-ffda54d85cc3aec3ba499c8c0eda9b2176edfe83/phpunit.xml
> 
> .............................................................   61 / 1352 (  4%)
> <snip>
> .............................................................  427 / 1352 ( 31%)
> .......................PHP Fatal error:  Allowed memory size of 134217728 bytes exhausted (tried to allocate 4096 bytes) in /dev/shm/BUILDROOT/php-twig-1.35.1-1.fc27.remi.7.1.x86_64/usr/share/php/Twig/Compiler.php on line 126
> 
> ```
> 
> P.S. twig test suite is only an example, this happens with lot of other projects
> 

Thank you very much for taking the time to test Snuffleupagus ♥
Unfortunately, I can't reproduce this behaviour here:

```
jvoisin@grimhilde 19:54 ~/dev/Twig phpunit -d
'extension=snuffleupagus.so' -d
'sp.configuration_file=/home/jvoisin/dev/snuffleupagus/config/default.rules'
PHPUnit 5.4.6 by Sebastian Bergmann and contributors.

Testing Twig Test Suite
.............................................................   61 /
1422 (  4%)
.............................................................  122 /
1422 (  8%)
.............................................................  183 /
1422 ( 12%)
.............................................................  244 /
1422 ( 17%)
.............................................................  305 /
1422 ( 21%)
.............................................................  366 /
1422 ( 25%)
.............................................................  427 /
1422 ( 30%)
........................S....................................  488 /
1422 ( 34%)
.............................................................  549 /
1422 ( 38%)
.............................................................  610 /
1422 ( 42%)
.............................................................  671 /
1422 ( 47%)
.............................................................  732 /
1422 ( 51%)
.............................................................  793 /
1422 ( 55%)
.............................................................  854 /
1422 ( 60%)
.............................................................  915 /
1422 ( 64%)
.............................................................  976 /
1422 ( 68%)
............................................................. 1037 /
1422 ( 72%)
............................................................. 1098 /
1422 ( 77%)
............................................................. 1159 /
1422 ( 81%)
............................................................. 1220 /
1422 ( 85%)
............................................................. 1281 /
1422 ( 90%)
............................................................. 1342 /
1422 ( 94%)
............................................................. 1403 /
1422 ( 98%)
...................                                           1422 /
1422 (100%)

Time: 978 ms, Memory: 26.00MB

OK, but incomplete, skipped, or risky tests!
Tests: 1422, Assertions: 3408, Skipped: 1.
jvoisin@grimhilde 19:55 ~/dev/Twig
```

How are you running the testsuite?

**Comment by jvoisin** [comment]
Thank you very much for taking the time to test Snuffleupagus ♥
Unfortunately, I can't reproduce this behaviour here:

```
jvoisin@grimhilde 19:54 ~/dev/Twig phpunit -d 'extension=snuffleupagus.so' -d 'sp.configuration_file=/home/jvoisin/dev/snuffleupagus/config/default.rules' PHPUnit 5.4.6 by Sebastian Bergmann and contributors.

Testing Twig Test Suite
.............................................................   61 /
1422 (  4%)
.............................................................  122 /
1422 (  8%)
.............................................................  183 /
1422 ( 12%)
.............................................................  244 /
1422 ( 17%)
.............................................................  305 /
1422 ( 21%)
.............................................................  366 /
1422 ( 25%)
.............................................................  427 /
1422 ( 30%)
........................S....................................  488 /
1422 ( 34%)
.............................................................  549 /
1422 ( 38%)
.............................................................  610 /
1422 ( 42%)
.............................................................  671 /
1422 ( 47%)
.............................................................  732 /
1422 ( 51%)
.............................................................  793 /
1422 ( 55%)
.............................................................  854 /
1422 ( 60%)
.............................................................  915 /
1422 ( 64%)
.............................................................  976 /
1422 ( 68%)
............................................................. 1037 /
1422 ( 72%)
............................................................. 1098 /
1422 ( 77%)
............................................................. 1159 /
1422 ( 81%)
............................................................. 1220 /
1422 ( 85%)
............................................................. 1281 /
1422 ( 90%)
............................................................. 1342 /
1422 ( 94%)
............................................................. 1403 /
1422 ( 98%)
...................                                           1422 /
1422 (100%)

Time: 978 ms, Memory: 26.00MB

OK, but incomplete, skipped, or risky tests!
Tests: 1422, Assertions: 3408, Skipped: 1.
jvoisin@grimhilde 19:55 ~/dev/Twig
```

How are you running the testsuite?

(It seems that github doesn't allow markdown in email replies </3)

**Comment by remicollet** [comment]
> How are you running the testsuite? 

-d option should be passed to php, not to phpunit

`php70 /usr/bin/phpunit -d extension=snuffleupagus.so` => extension is not enabled, so test suite passes

`php70  -d extension=snuffleupagus.so /usr/bin/phpunit ` => extension is enabled, test suite goes OOM

**Comment by jvoisin** [comment]
Even so, I still can't reproduce :/

```
jvoisin@mim 8:59 ~/Dev/Twig php7.1 -d extension=/home/jvoisin/Dev/snuffleupagus/src/modules/snuffleupagus.so /usr/bin/phpunit                                                                         [2.x] git:Twig
PHPUnit 5.4.6 by Sebastian Bergmann and contributors.

Testing Twig Test Suite
.............................................................   61 / 1424 (  4%)
.............................................................  122 / 1424 (  8%)
.............................................................  183 / 1424 ( 12%)
.............................................................  244 / 1424 ( 17%)
.............................................................  305 / 1424 ( 21%)
.............................................................  366 / 1424 ( 25%)
.............................................................  427 / 1424 ( 29%)
..........................S..................................  488 / 1424 ( 34%)
.............................................................  549 / 1424 ( 38%)
.............................................................  610 / 1424 ( 42%)
.............................................................  671 / 1424 ( 47%)
.............................................................  732 / 1424 ( 51%)
.............................................................  793 / 1424 ( 55%)
.............................................................  854 / 1424 ( 59%)
.............................................................  915 / 1424 ( 64%)
.............................................................  976 / 1424 ( 68%)
............................................................. 1037 / 1424 ( 72%)
............................................................. 1098 / 1424 ( 77%)
............................................................. 1159 / 1424 ( 81%)
............................................................. 1220 / 1424 ( 85%)
............................................................. 1281 / 1424 ( 89%)
............................................................. 1342 / 1424 ( 94%)
............................................................. 1403 / 1424 ( 98%)
.....................                                         1424 / 1424 (100%)

Time: 838 ms, Memory: 24.00MB

OK, but incomplete, skipped, or risky tests!
Tests: 1424, Assertions: 3411, Skipped: 1.
jvoisin@mim 8:59 ~/Dev/Twig file /home/jvoisin/Dev/snuffleupagus/src/modules/snuffleupagus.so                                                                                                         [2.x] git:Twig
/home/jvoisin/Dev/snuffleupagus/src/modules/snuffleupagus.so: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, BuildID[sha1]=5fb20866bf18b30561bda515bb11230decf7a100, with debug_info, not stripped
```

**Comment by remicollet** [comment]
This looks like it only fails when `sp.configuration_file` is set

In 40-snuffleupagus.ini
`extension = snuffleupagus.so`
=> Test suite passes

In 40-snuffleupagus.ini
```
extension = snuffleupagus.so
sp.configuration_file = '/etc/php.d/snuffleupagus-*.rules'
```
And with a single `snuffleupagus-default.rules` file which only have
`# Default rules see https://snuffleupagus.readthedocs.io/config.html
`
Then, test suite goes OOM

**Comment by remicollet** [comment]
See PR #161  which fix this issue

**Comment by remicollet** [comment]
So already fixed by 309481168de02f2dee5a4266359d72866442f665

**Comment by jvoisin** [comment]
So, you tested Snuffleupagus, found a bug, spent time find the root cause, and submitting a pull-request to fix it? You're a wonderful person ♥

We will be happy, to offer your some beers/beverages of your choice if we ever meet you AFK ♥
